Add async support

pauloromeira opened the pull request:
Implements: https://github.com/fabfuel/circuitbreaker/issues/4

Hi @fabfuel ,

First of all, congrats for the lib!

This PR adds `async` support to it, but this change will drop python 2 support and some python 3 early versions.

I'm opening it now to kindly ask you some help for making this available for the majority of python versions and change setup / gh actions accordingly.

Thanks in advance!

fabfuel commented:
Hi @pauloromeira, thanks for the PR!
What versions do you aim to support in the end? 
Currently tests only pass on 3.10, but this is only used by 16% of the users:

<img width="383" alt="Bildschirmfoto 2023-04-12 um 06 59 09" src="https://user-images.githubusercontent.com/1582291/231354279-490c4cbc-80eb-4182-88cb-e21c491b6d78.png">

Best
Fabian

pauloromeira commented:
Hi @fabfuel, thanks for the reply and the interesting usage information! I'll start fixing issues for `Python >= 3.6`.

I believe there are two options:
1. Support `Python >= 3.5`: + async functions ([PEP-0492](https://peps.python.org/pep-0492/))
2. Support `Python >= 3.6`: + async functions + async generators ([PEP-0525](https://peps.python.org/pep-0525/))

Based on the numbers you gave, I believe we could include `async generators` as well (`>= 3.6`). First because `3.5` usage is below 2%, second because it's not a huge leap if people want to upgrade from `Python 3.5` to `Python 3.6`.

What do you think? 

PS: I also check the numbers starting from the beginning of this year. The percentages are very similar:
![image](https://user-images.githubusercontent.com/1191135/231529919-057e724b-f713-4588-8c85-3c160d6a07b1.png)

Best,
Paulo

pauloromeira commented:
Hi @fabfuel!

The code is now passing for: `3.7`, `3.8`, `3.9`, `3.10`, `3.11`. Maybe it works also for `Python 3.6` but unfortunately [pytest-asyncio](https://github.com/pytest-dev/pytest-asyncio) doesn't support `3.6` which is required for pytest to run async tests.

I know you might not like dropping support for versions `2.7`, `3.5` and `3.6`, but I believe usage can increase with async support. Specially since [pybreaker](https://github.com/danielfm/pybreaker), the most popular alternative, is not supporting async yet.

Two notes:
1. I slightly changed the `fallback` functionality for `generators` to make everything more consistent. Before it was possible to have a fallback function that was **not** a generator. Well, if the fallback function is called implicitly, I believe it should have the **same interface** as the decorated function, not just its args. So I changed that for every decorated function: if `sync`, the fallback must be `sync`; if `generator`, the fallback must be a `generator`; if `async`, the fallback must be `async` and if `async generator`, the fallback must be `async generator` - I added this information to the README.

2. The tests are now running for every supported function types: `sync-function`, `sync-generator`, `async-function` and `async-generator`. Wherever you see the `resolve_call` fixture, all of these function types are being tested. This is not only extending generators tests but reusing almost all existing tests for the asynchronous versions.

I hope you like and decide to merge it! 

Regards,
Paulo

fabfuel commented:
Thanks a lot @pauloromeira! Sorry, I'm a bit behind as i had no time to fully review yet, but what I saw so far looks really good. I try to find time to also test it as soon as possible! 🙂

fabfuel reviewed (approved):
Sorry for the long delay @pauloromeira, I needed to prepare for parental leave and had very busy last weeks while currently not working in the Python universe, so I needed some time to get into the topic again. I also did not use Python async functions in production yet, so it's awesome, that you took care for adding the functionality 🙏 

Thanks a lot for your PR, it's really well done. I highly appreciate your effort and the way how you integrated the async functionality into the existing library in an unobtrusive way (=without rewriting everything 😉).

The tests have been quite rewritten though, so I needed some time to test myself and get back into it. But finally I reviewed it and I'm happy to merge.

I will create version 2.0, I think this new ability deserves a new major version, especially also as we drop Python2 support.

Thanks a lot @pauloromeira!

pauloromeira commented:
You're welcome, @fabfuel! Thank you for dedicating your time and effort to reviewing it. I appreciate your decision to proceed with the merge, despite the drop of support for Python 2.

Wishing you all the best in your journey of parenthood! 🎉
